engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala

```diff
@@ -0,0 +1,142 @@
+      }
+    })
+
+    partialSourcesToSourceCollections(partialSources.tryToErrorOr, allowNoInputs).errorOrToTry
+  }
+
+  private def workflowInputs(data: String): Vector[WdlJson] = {
+    import spray.json._
+    data.parseJson match {
+      case JsArray(Seq(x, xs@_*)) => (Vector(x) ++ xs).map(_.compactPrint)
+      case JsArray(_) => Vector.empty
+      case v: JsValue => Vector(v.compactPrint)
+    }
+  }
+
+  private def partialSourcesToSourceCollections(partialSources: ErrorOr[PartialWorkflowSources], allowNoInputs: Boolean): ErrorOr[Seq[WorkflowSourceFilesCollection]] = {
+    def validateInputs(pws: PartialWorkflowSources): ErrorOr[Seq[WdlJson]] =
+      (pws.workflowInputs.isEmpty, allowNoInputs) match {
+        case (true, true) => Vector("{}").validNel
+        case (true, false) => "No inputs were provided".invalidNel
+        case _ =>
+          val sortedInputAuxes = pws.workflowInputsAux.toSeq.sortBy { case (index, _) => index } map { case(_, inputJson) => Option(inputJson) }
+          (pws.workflowInputs map { workflowInputSet: WdlJson => mergeMaps(Seq(Option(workflowInputSet)) ++ sortedInputAuxes).toString }).validNel
+      }
+
+    def validateOptions(options: Option[WorkflowOptionsJson]): ErrorOr[WorkflowOptions] =
+      WorkflowOptions.fromJsonString(options.getOrElse("{}")).tryToErrorOr leftMap { _ map { i => s"Invalid workflow options provided: $i" } }
+
+    def validateWorkflowSource(partialSource: PartialWorkflowSources): ErrorOr[WdlJson] = partialSource.workflowSource match {
+      case Some(src) => src.validNel
+      case _ => s"Incomplete workflow submission: $partialSource".invalidNel
+    }
+
+    partialSources match {
+      case Valid(partialSource) =>
+        (validateWorkflowSource(partialSource) |@| validateInputs(partialSource) |@| validateOptions(partialSource.workflowOptions)) map {
+          case (wfSource, wfInputs, wfOptions) =>
+            wfInputs.map(inputsJson => WorkflowSourceFilesCollection(
+              workflowSource = wfSource,
+              workflowType = partialSource.workflowType,
+              workflowTypeVersion = partialSource.workflowTypeVersion,
+              inputsJson = inputsJson,
+              workflowOptionsJson = wfOptions.asPrettyJson,
+              labelsJson = partialSource.customLabels.getOrElse("{}"),
+              importsFile = partialSource.zippedImports))        }
+      case Invalid(err) => err.invalid
+    }
+  }
+
+  private def deprecationWarning(out: String, in: String): Unit = {
+    val warning =
+      s"""
+         |The '$out' parameter name has been deprecated in favor of '$in'.
+         |Support for '$out' will be removed from future versions of Cromwell.
+         |Please switch to using '$in' in future submissions.
+         """.stripMargin
+    log.warn(warning)
+  }
+
+  def mergeMaps(allInputs: Seq[Option[String]]): JsObject = {
+    val convertToMap = allInputs.map(x => toMap(x))
+    JsObject(convertToMap reduce (_ ++ _))
```

Comment:
I think we might want `reduceLeft` here, otherwise the order is [nondeterministic](https://www.scala-lang.org/api/current/scala/collection/immutable/List.html#reduce[A1>:A](op:(A1,A1)=>A1):A1)